pkg/common/moerr/error.go

```diff
@@ -57,24 +63,118 @@ const (
+	// ErrUnresolvedConflict read transaction encounters unresolved data
+	ErrUnresolvedConflict = 10003
 	// ErrTxnError TxnError wrapper
 	ErrTxnError = 10004
 	// ErrDNShardNotFound DNShard not found, need to get the latest DN list from HAKeeper
 	ErrDNShardNotFound = 10005
+
+	// ErrEnd, the max value of MOErrorCode
+	ErrEnd = 65535
 )
 
+type moErrorMsgItem struct {
+	errorCode        MOErrorCode
+	mysqlErrorCode   uint16
+	errorMsgOrFormat string
+}
+
+var errorMsgRefer = map[int32]moErrorMsgItem{
+	SUCCESS: {0, 0, "ok"},
+	INFO:    {20001, 0, "%s"},
+	WARN:    {20002, 0, "%s"},
+
+	// Group 1: Internal errors
+	ERROR_START:              {21000, 0, "%s"},
+	INTERNAL_ERROR:           {21001, 0, "%s"},
+	NYI:                      {21002, 0, "%s"},
+	ERROR_FUNCTION_PARAMETER: {21003, 0, "%s"},
+
+	// Group 2: numeric
+	DIVIVISION_BY_ZERO: {22000, 0, "division by zero"},
+	OUT_OF_RANGE:       {22001, 0, "overflow from %s to %s"},
+	DATA_TRUNCATED:     {22002, 0, "%s data truncated"},
+	INVALID_ARGUMENT:   {22003, 0, "%s"},
+
+	// Group 3: invalid input
+	BAD_CONFIGURATION: {23000, 0, "invalid %s configuration"},
+	INVALID_INPUT:     {23001, 0, "%s"},
+
+	// Group 4: unexpected state
+	INVALID_STATE:         {24000, 0, "%s"},
+	LOG_SERVICE_NOT_READY: {24001, 0, "log service not ready"},
+
+	// group 5: rpc timeout
+	ErrRPCTimeout:         {25000, 0, "%s"},
+	ErrClientClosed:       {25001, 0, "client closed"},
+	ErrBackendClosed:      {25002, 0, "backend closed"},
+	ErrStreamClosed:       {25003, 0, "stream closed"},
+	ErrNoAvailableBackend: {25004, 0, "no available backend"},
+
+	// Group 1: sql error
+	ErrNoDatabaseSelected: {26000, 1046, "No database selected"},
+
+	// Group 10: txn
+	ErrTxnClosed:          {30000, 0, "the transaction has been committed or aborted"},
+	ErrTxnWriteConflict:   {30001, 0, "write conflict"},
+	ErrMissingTxn:         {30002, 0, "missing txn"},
+	ErrUnresolvedConflict: {30003, 0, "unresolved conflict"},
+	ErrTxnError:           {30004, 0, "%s"},
+	ErrDNShardNotFound:    {30005, 0, "%s"},
+
+	// Group End: max value of MOErrorCode
+	ErrEnd: {65535, 0, "%s"},
+}
+
+func New(code int32, args ...any) *Error {
+	return newWithDepth(Context(), 1, code, args...)
+}
+
+func NewWithContext(ctx context.Context, code int32, args ...any) *Error {
+	return newWithDepth(ctx, 1, code, args...)
+}
+
+func newWithDepth(ctx context.Context, depth int, code int32, args ...any) *Error {
+	var err *Error
+	item, has := errorMsgRefer[code]
+	if !has {
+		panic(fmt.Errorf("not exist MOErrorCode: %d", code))
+	}
+	if len(args) == 0 {
+		err = &Error{
+			MysqlErrCode: item.mysqlErrorCode,
+			Code:         code,
+			ErrorCode:    item.errorCode,
+			Message:      item.errorMsgOrFormat,
+		}
+	} else {
+		err = &Error{
+			MysqlErrCode: item.mysqlErrorCode,
+			Code:         code,
+			ErrorCode:    item.errorCode,
+			Message:      fmt.Sprintf(item.errorMsgOrFormat, args...),
+		}
+	}
+	_ = errors.WithContextWithDepth(ctx, err, depth+1)
```

Comment:
Minor, will we ever call newWithDepth with depth not equals to 1?

Reply:
got it, obviously no.